### PR TITLE
Convert Python Bazel targets to Make targets

### LINKED ISSUE #23830: Build and test Python without bazel
**What would you like to be added**:

Build and test Python without bazel, the related bazel rules are:

- py_binary
- py_test

Things to be considered:

- This will need to co-exist with current bazel rules for a short while, delete Python test targets from bazel once this confirmed working
- Test driven by make rule and bash scripts
- Ensure hermetic testing, including fixed tools version, like python, python3, dependencies
- Test results are in junit results format

**Why is this needed**:

This is part of #23796 

**Recommended testing approach**:
- Make test invokable by `make test`
- Create a PullRequest with change, manually run job by commenting `/test pull-test-infra-test-beta`
- [Optional]: if there is any tool that needs to exist in test image, update image at https://github.com/kubernetes/test-infra/blob/6e0d21ee50d690c8f78239257f588021d2ed57b5/config/jobs/kubernetes/test-infra/test-infra-presubmits.yaml#L145

/sig testing
/help

### PR DESCRIPTION
**Pull request by khalludi**
Converts the Python Bazel targets to make targets, placing binaries in a folder called _bin_make, and runs all the python test targets when `make test` is run in top-level directory.

Resolves #23830

### CODE REVIEW

**Comment by khalludi** [comment]
Hey, when I double-checked the issue, I saw that you already had a test image defined here: `test-infra/config/jobs/kubernetes/test-infra/test-infra-presubmits.yaml`. I thought this meant that it would run the `make test` to run all tests, so I didn't need to create a custom image. However, I can create a Dockerfile with the required Python dependencies that run the `make test` command.

Something I wasn't clear on was building and running the Python targets. At the moment I haven't really included targets for running the binaries. I assume that we want the users to run the files in their respective project directories, so there shouldn't be a way of invoking them from the root directory. For example, if I want to run `make prepare_release_branch`, I need to do this inside `releng`, not inside `test-infra`.

For building, I didn't include a `make build` command at the top level to build all the python targets. Is this a feature we want to include? 

You'll also notice that I do have some build commands in the Makefiles that basically move the python files into a top-level directory `_bin_make` that is created on the fly, mimicking bazel compile behavior. Do we want to do this or should we just not have a build command since these are Python files and are executable as is? 

Note that no file in a project directory is being shared or used in another project, so not sure if it is even necessary to store "executables" at the top level. Rather, maybe we should have a bin created inside the necessary project directory.

**Comment by khalludi** [comment]
Gotcha. I'll have to do a bit of digging on what prow is but that makes sense. I will attempt to implement it tonight.

**Comment by khalludi** [comment]
So, I looked a little bit into the prow jobs but I'm a bit lost. As I understand, I have a Dockerfile that needs to be ran, build the specified image, and push onto the GCR. I found link 1 that has jobs to push images. Scrolling down the `README.md`, I see the code for `Simple Build Example`. I think I'm supposed to create that yaml file in the `image-pushing/` directory with only the docker build step and changing the args to point to my Dockerfile. Is this correct?

If so, I won't be able to copy the `requirements3.txt` file into the image, resulting in me hardcoding the list into the Dockerfile. This is because I don't see how I can pass arguments to my Dockerfile that is being built with another image.

Sorry, I'm just trying to understand what I'm doing before I do it. Thanks.

[1] https://github.com/kubernetes/test-infra/tree/master/config/jobs/image-pushing

**Comment by khalludi** [comment]
So, I think I added the right changes. In the `staging-testinfra.yaml`, I added the `push-pytest` job which should build and push the image onto GCR. Then, I moved the `Dockerfile` to `images/pytest`. Inside, I am copying the `requirements3.txt` and installing the necessary dependencies. I also added a `cloudbuild.yaml` for building the image.

I'm not sure if the image name will be correct though when doing the `make test` because I don't know what the `$PROJECT_ID` or `$GIT_TAG` will be. I kept the `pytests` target so that the Docker image could run it.

I heavily followed the examples. Does everything look good?

**Comment by chaodaiG** [comment]
@ khalludi, still on track with this PR?

**Comment by khalludi** [comment]
Sorry, life got busy. I plan to get to this by Wednesday night. Thanks for the patience!

File: config/jobs/image-pushing/k8s-staging-test-infra.yaml [hunk 1]
```diff
@@ -273,3 +273,27 @@ postsubmits:
           - --project=k8s-staging-test-infra
           - --env-passthrough=PULL_BASE_REF
           - triage/
+    - name: post-test-infra-push-pytest
```
[the remark this answers is not in the document]
Comment (khalludi) [reply config/jobs/image-pushing/k8s-staging-test-infra.yaml]: Changes were moved and merged in #24235.

File: images/pytest/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,12 @@
+FROM python:3.8-slim-buster
```
[the remark this answers is not in the document]
Comment (khalludi) [reply images/pytest/Dockerfile]: Moved and merged in #24235

File: images/pytest/cloudbuild.yaml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+steps:
```
[the remark this answers is not in the document]
Comment (khalludi) [reply images/pytest/cloudbuild.yaml]: Moved and merged in #24235

**Comment by khalludi** [comment]
So, I made the necessary changes and rebased. All tests are passing. Please let me know what the next steps are @chaodaiG. I have the next week off and plan to spend time finishing this PR.

Thanks.

File: Makefile [hunk 4]
```diff
@@ -29,15 +29,48 @@ export PATH GOROOT GO111MODULE
 SPACE:=$(subst ,, )
 SHELL:=env PATH=$(subst $(SPACE),\$(SPACE),$(PATH)) $(SHELL)
 ################################################################################
+# ================================ Building ===================================
+DOCKER_IMAGE_NAME=pytest-infra
+
+docker-build:
+	docker build -t $(DOCKER_IMAGE_NAME) .
```
Comment (chaodaiG) [inline Makefile:36]: Not seeing Dockerfile defined, so this might fail?

**Review by chaodaiG — COMMENTED** [review]
If I understand correctly, the current implementation is building a docker image every time before running python tests?

Also, seems like `make -C ./metrics test` still runs locally, I think we would prefer consistent solution across, in another word, `make -C ./metrics test` runs metrics test in a docker container, and `make pytest` runs all tests in one or more docker container(s)

**Comment by chaodaiG** [comment]
@khalludi , please take a look at https://github.com/kubernetes/test-infra/pull/24491, I think the idea there is what we might want to consider

**Comment by chaodaiG** [comment]
@khalludi , still on track working on this PR? Please let me know what I can help

**Comment by chaodaiG** [comment]
This has been pending for 3 weeks. I have done what I could do on the other front, and we are unfortunately running some of the tests duplicated in bazel and make rules, which won't be resolved until fully deprecation of bazel, so would be nice if this could be done sooner than later.

@khalludi , thank you for the contribution so far! And I understand that you might be busy with your day job, and that's totally acceptable. I'm planning to pick this PR up from here starting next year, please respond here and let me know what you think.

**Comment by khalludi** [comment]
Hi Chao, I apologize for the silence. I just haven't been finding the time to continue working on this PR. I am fine with you picking this up in the next year. Thank you for all the help, I have learned a lot!

**Comment by chaodaiG** [comment]
Thank you @khalludi !

**Comment by chaodaiG** [comment]
/hold
I'll iterate on this work by creating a separate PR

**Comment by chaodaiG** [comment]
I've squashed all commits from your branch and created https://github.com/kubernetes/test-infra/pull/24664

**Comment by chaodaiG** [comment]
the work from this PR was cherry-picked and merged as part of https://github.com/kubernetes/test-infra/issues/23830
/close

**Pull request closed without merging**